analyzers/src/SonarAnalyzer.CSharp/Rules/SillyBitwiseOperation.cs

```diff
@@ -64,5 +84,101 @@ private void CheckBinary(SonarSyntaxNodeReportingContext context, int constValue
             var binary = (BinaryExpressionSyntax)context.Node;
             CheckBinary(context, binary.Left, binary.OperatorToken, binary.Right, constValueToLookFor);
         }
+
+        private void CheckComparison(SonarSyntaxNodeReportingContext context)
+        {
+            var binary = (BinaryExpressionSyntax)context.Node;
+            if (ShouldRaise(context.SemanticModel, binary.Left, binary.Right, binary.Kind(), false))
+            {
+                context.ReportIssue(Diagnostic.Create(BitwiseRule, binary.GetLocation()));
+            }
+
+            if (ShouldRaise(context.SemanticModel, binary.Right, binary.Left, binary.Kind(), true))
+            {
+                context.ReportIssue(Diagnostic.Create(BitwiseRule, binary.GetLocation()));
+            }
+        }
+
+        private bool ShouldRaise(SemanticModel semanticModel, SyntaxNode value, SyntaxNode expectedConstant, SyntaxKind kind, bool constantIsLeft)
```

Comment:
Letsonly implement the out of range check. You only need the binary for that.
```suggestion
        private bool ShouldRaise(SemanticModel semanticModel, SyntaxNode binary)
```

You need the following ingredients:
* `bool TryGetConstant(SemanticModel model, SyntaxNode binary, out object constantValue, out SyntaxNode opposite)` -> returns true, if Left or Right of binary is a constant and sets constantValue to the constant and opposite to the Left or Right node of the binary opposite to the constant.
* `semanticModel.GetType(opposite)` get the type of the target expression.
* `bool TryRangeOfType(ITypeSymbol type, out double min, out double max)`
* convert constantValue to double
* Check if constantValue is outside the range:
  * true: is target type whitelisted (float, half, double)? true -> report, else -> done
  * false: We need to check the boundary also. But not now.

Everything else is not needed at the moment.

The boundary checks will be implemented in a second step or even a follow up PR.

Reply:
I refactored it to only implement out-of-range checks for floats (no edge-cases, no other types).

My implementation follows your suggestion roughly 90%, with only one main different:
>  returns true, if Left or Right of binary is a constant and sets constantValue to the constant and opposite to the Left or Right node of the binary opposite to the constant.

I wrote this slightly different, after some trial and error, to pass in first left-right and then right-left, since it's possible that both of the operands are "constants". 

For example:
```csharp
float x = 42;
const double y = double.MaxValue;
_ = x < y;
```
In this case, the semantic model can deduce that x has a "constant value", since it can reason easily what the value is.
Handling the example:

```csharp
// _ = x < y;
if (IsConstant(left)) // 42, constant, but does not overflow. Does not raise.
{
// do work
}
else if (IsConstant(right)) // double.MaxValue, constant, overflows "float", but this is never reached because it's on the else.
{
// do work
}
```

That is why my method's signature does not look like:

`bool TryGetConstant(SemanticModel model, SyntaxNode binary, out object constantValue, out SyntaxNode opposite)`

But like:
`bool TryGetConstant(SemanticModel model, SyntaxNode expectedConstant, out double constant)`